src/v/storage/parser.cc

```diff
@@ -245,4 +252,89 @@ ss::future<result<size_t>> continuous_batch_parser::consume() {
           return result<size_t>(_err);
       });
 }
+
+class copy_helper {
+public:
+    explicit copy_helper(
+      ss::input_stream<char> input,
+      ss::output_stream<char> output,
+      record_batch_transform_predicate pred)
+      : _input(std::move(input))
+      , _output(std::move(output))
+      , _pred(std::move(pred)) {}
+
+    ss::future<result<model::record_batch_header>> read_header() {
+        return read_header_impl(_input, ss::sstring("copy_helper"));
+    }
+
+    /// Copy data.
+    /// Return number of bytes copied.
+    ss::future<result<size_t>> run() {
+        size_t consumed = 0;
+        bool stop = false;
+        while (!stop) {
+            auto r = co_await read_header();
+            if (!r) {
+                if (r.error() == parser_errc::end_of_stream) {
+                    break;
+                }
+                co_return r.error();
+            }
+
+            _header = r.value();
+            auto remaining = _header.size_bytes
+                             - model::packed_record_batch_header_size;
+
+            // invoke pred
+            switch (_pred(_header)) {
+            case batch_consumer::consume_result::skip_batch:
+                // Record batch shouldn't be copied, we can just skip its
+                // content altogether
+                co_await _input.skip(remaining);
+                break;
+            case batch_consumer::consume_result::accept_batch: {
+                auto body = co_await verify_read_iobuf(
+                  _input, remaining, "copy_helper");
+                if (!body) {
+                    co_return body.error();
+                }
+                // we should only write to the output stream if we can
+                // guarantee that both header and records are available
+
+                // the header should be re-serialized since the predicate
+                // might change it in-place (this is a low level tool)
+                // we're also need to update header only crc
+                _header.header_crc = model::internal_header_only_crc(_header);
+                iobuf hdr = disk_header_to_iobuf(_header);
+                co_await write_iobuf_to_output_stream(std::move(hdr), _output);
+                co_await write_iobuf_to_output_stream(
+                  std::move(body.value()), _output);
+                consumed += _header.size_bytes;
+                break;
+            }
+            case batch_consumer::consume_result::stop_parser:
+                stop = true;
+                break;
```

Comment:
Can we implement the `copy_header` as an implementation of `batch_consumer` or a `segment_reader` so that we don't accumulate utilities within the parser? If not then perhaps we need to enhance the interface.

Reply:
@Lazin i'm not really sure why you would need to futurize the batch_consumer. for example the log reader runs on top of the batch consumer. if you add a record batch reader it will probably solve the issue.